### ISSUE TITLE
Auth: Custom strategy scheme does not resolve

### ISSUE THREAD

**Issue by Caceresenzo** [Teranode/nuxt-module-alternatives#56]
Hello

I am using [a1ter's nuxt-auth-aws-cognito-scheme](https://www.npmjs.com/package/@a1ter/nuxt-auth-aws-cognito-scheme) package, but its not working.

```
 ERROR  Cannot start nuxt:  Cannot read properties of undefined (reading 'as')                                                                                                                        18:37:09

  at /DIR/webapp/node_modules/@nuxtjs-alt/auth/dist/module.mjs:614:32
  at Array.filter (<anonymous>)
  at setup (/DIR/webapp/node_modules/@nuxtjs-alt/auth/dist/module.mjs:612:57)    <-- line is not valid since I was debugging stuff
  at Object.normalizedModule (/DIR/webapp/node_modules/@nuxt/kit/dist/index.mjs:607:29)
  at async installModule (/DIR/webapp/node_modules/@nuxt/kit/dist/index.mjs:432:3)
  at async initNuxt (/DIR/webapp/node_modules/nuxt/dist/index.mjs:1423:7)
  at async load (/DIR/webapp/node_modules/nuxi/dist/chunks/dev.mjs:6750:9)
  at async Object.invoke (/DIR/webapp/node_modules/nuxi/dist/chunks/dev.mjs:6795:5)
  at async _main (/DIR/webapp/node_modules/nuxi/dist/cli.mjs:46:20)
```

Here is the config I was using:

```typescript
import { defineNuxtConfig } from 'nuxt'

// https://v3.nuxtjs.org/api/configuration/nuxt.config
export default defineNuxtConfig({
  auth: {
    strategies: {
      cognito: {
        enabled: true,
        name: "cognito",
        scheme: '@a1ter/nuxt-auth-aws-cognito-scheme/scheme/scheme',
        credentials: {
          userPoolId: "...",
          userPoolWebClientId: "...",
          region: "..."
        },
        endpoints: {
          user: false,
        },
      },
    }
  }
})
```

The problem seems to lies here:
https://github.com/Teranode/nuxt-module-alternatives/blob/4e36664ecc3f75bf64681607d62787130911143b/%40nuxtjs-alt/auth/src/module.ts#L40-L44

You are calling `resolveStrategies` which fails to resolve the package:
https://github.com/Teranode/nuxt-module-alternatives/blob/4e36664ecc3f75bf64681607d62787130911143b/%40nuxtjs-alt/auth/src/module.ts#L35

And after a bit of `console.log`, I found out that
https://github.com/Teranode/nuxt-module-alternatives/blob/4e36664ecc3f75bf64681607d62787130911143b/%40nuxtjs-alt/auth/src/resolve.ts#L95
Isn't working as expected since `resolvePath` is returning a promise! And so `existsSync` always return `false`.

I tried to add some `async/await` up to the `setup` method, but since I was working with a `dist` version, it was not working for me in the end.... But the problem of non resolving scheme was fixed!

**Comment by Teranode** [comment]
@Caceresenzo 
Published an npm update.

But a couple of issues will arrive from this.

1. Since this is nuxt3, importing the module will have to work a bit differently. When specifying the strategy, you will have to name the strategy as the module you're importing:

e.g (You do not need to specify `name` or `scheme` if you follow what i said below, this is just an example if it's not there in the module)

```ts
cognito: {
    name: '@sirdiego/nuxt-auth-cognito-scheme', // you can either use 'name' or 'provider'
    scheme: '@sirdiego/nuxt-auth-cognito-scheme/CognitoAuthScheme',
    tokenType: "Bearer",
    globalToken: true,
    tokenRequired: true,
    tokenName: "Authorization",
    autoFetchUser: true,
    userPoolId: process.env.AWS_COGNITO_USER_POOL_ID,
    clientId: process.env.AWS_COGNITO_CLIENT_ID,
    refreshInterval: 5 * 60 * 1000,
    fetchUserCallback: false
},
```

2. The module you are trying to import only works for nuxt 2 (and the global fetch require is messing with the import and isnt needed since nuxt comes with fetch).

3. That being said, if you're okay with being hacky you can edit the [nuxt-auth-aws-cognito-scheme](https://www.npmjs.com/package/@a1ter/nuxt-auth-aws-cognito-scheme) package and change `index.js` to:

```ts
import { name, version } from './package.json'
import { defineNuxtModule } from '@nuxt/kit';

export default defineNuxtModule({
    meta: {
        name,
        version,
        configKey: 'cognito',
        compatibility: {
            nuxt: "^3.0.0",
        },
    },
    async setup(moduleOptions, nuxt) {
        nuxt.options.build.transpile.push("@sirdiego/nuxt-auth-cognito-scheme");

        nuxt.options.auth = nuxt.options.auth || {};
        nuxt.options.auth.strategies = nuxt.options.auth.strategies || {};
        nuxt.options.auth.strategies.cognito = nuxt.options.auth.strategies.cognito || {};
        nuxt.options.auth.strategies.cognito.name = "@sirdiego/nuxt-auth-cognito-scheme";
        nuxt.options.auth.strategies.cognito.scheme = "@sirdiego/nuxt-auth-cognito-scheme/CognitoAuthScheme";
    },
});
```

and `CognitoAuthScheme.js` remove: 
```ts
if (!process.client || !"fetch" in window) {
  require("cross-fetch/polyfill");
}
```

Then finally add the module after the auth module (nuxt3 sorts by ascending so put it above the auth module)

It should work then. I would provide an actual module, but im still deciding if to discontinue the module when the official auth module decides to release.

**Comment by Teranode** [comment]
The specific package you're using has an issue with vite.

**Comment by Caceresenzo** [comment]
Thanks a lot man!
I wasn't expecting this level of modification.

I am still doing research on what stack to use and I cannot lose too much time on this...

I will test it when I have time!

**Comment by Teranode** [comment]
If the solution above isn't working for you, and you prefer `@a1ter/nuxt-auth-aws-cognito-scheme` I can take a look at it sometime later.

**Comment by Caceresenzo** [comment]
Its not that I can't work with your solution, its that Cognito is using a hosted UI to do his things normally.

And I discovered only after creating the issue that using the custom scheme was working *but only for login*.

If you try to log to an account that didn't changed his temporary password or isn't verified, an error is raised and would be difficult to continue since we do not have access to the underlying Amplify instance (that the custom scheme is using).

I am still searching for better solution.
I created this issue issue so you would be able to resolve the not resoving `scheme` issue.
